File: core-java/src/test/java/chessgame/domain/model/QueenWhiteTest.java
package chessgame.domain.model;

import chessgame.domain.Piece;
import org.junit.Before;
import org.junit.Test;

import java.util.Set;

import static chessgame.domain.factory.PieceFactory.createWhiteQueen;
import static chessgame.domain.model.BoardPosition.A1;
import static chessgame.domain.model.BoardPosition.A8;
import static chessgame.domain.model.BoardPosition.B1;
import static chessgame.domain.model.BoardPosition.B3;
import static chessgame.domain.model.BoardPosition.D1;
import static chessgame.domain.model.BoardPosition.H1;
import static chessgame.domain.model.BoardPosition.H8;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

public class QueenWhiteTest {

    private static final BoardPosition INITIAL_POSITION = D1;

    private Board board;

    private Piece whiteQueen;

    @Before
    public void setUp() {
        board = new Board();
        board.init();
        whiteQueen = createWhiteQueen(board);
    }

    @Test
    public void createInstanceTest() {
        assertTrue(whiteQueen instanceof Queen);
    }

    @Test
    public void initPositionTest() {
        assertEquals(INITIAL_POSITION, whiteQueen.getPosition());
        assertEquals(whiteQueen, board.get(INITIAL_POSITION));
    }

    @Test
    public void moveToA1Test() {
        assertNotNull(board.get(INITIAL_POSITION.toString()));
        assertNull(board.get(A1));

        whiteQueen.moveTo(A1);

        assertNull(board.get(INITIAL_POSITION));
        assertNotNull(board.get("A1"));
    }

    @Test
    public void getAvailablePositionsFromD5Test() {
        final Set<BoardPosition> pos = whiteQueen
                .moveTo("D5")
                .getAvailablePositions();
        assertEquals(27, pos.size());
    }

    @Test
    public void getAvailablePositionsFromB4Test() {
        final Set<BoardPosition> pos = whiteQueen
                .moveTo(B1)
                .moveTo(B3)
                .getAvailablePositions();
        assertEquals(23, pos.size());
    }

    @Test
    public void getAvailablePositionsFromA1Test() {
        final Set<BoardPosition> pos = whiteQueen
                .moveTo(A1)
                .getAvailablePositions();
        assertEquals(21, pos.size());
    }

    @Test
    public void getAvailablePositionsFromH1Test() {
        final Set<BoardPosition> pos = whiteQueen
                .moveTo(H1)
                .getAvailablePositions();
        assertEquals(21, pos.size());
    }

    @Test
    public void getAvailablePositionsFromA8Test() {
        final Set<BoardPosition> pos = whiteQueen
                .moveTo(A1)
                .moveTo(A8)
                .getAvailablePositions();
        assertEquals(21, pos.size());
    }

    @Test
    public void getAvailablePositionsFromH8Test() {
        final Set<BoardPosition> pos = whiteQueen
                .moveTo(H1)
                .moveTo(H8)
                .getAvailablePositions();
        assertEquals(21, pos.size());
    }
}

File: core-kotlin/src/test/kotlin/chessgame/domain/model/QueenWhiteTest.kt
package chessgame.domain.model

import chessgame.domain.Piece
import chessgame.domain.PieceFactory.Companion.createWhiteQueen
import chessgame.domain.model.BoardPosition.*
import chessgame.domain.model.PieceColor.WHITE
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test

class QueenWhiteTest {

    companion object {
        private val INITIAL_POSITION = D1
    }

    private var board: Board? = null
    private var whiteQueen: Piece? = null

    @Before
    fun setUp() {
        board = Board()
        board!!.init()
        whiteQueen = createWhiteQueen(board!!)
    }

    @Test
    fun createInstanceTest() = assertTrue(whiteQueen is Queen)

    @Test
    fun whiteColorTest() = assertEquals(WHITE, whiteQueen!!.color)

    @Test
    fun initPositionTest() {
        assertEquals(INITIAL_POSITION, whiteQueen!!.position)
        assertEquals(whiteQueen, board!!.get(INITIAL_POSITION))
    }

    @Test
    fun moveToA1Test() {
        assertNotNull(board!![INITIAL_POSITION.toString()])
        assertNull(board!![A1])

        whiteQueen!!.moveTo(A1)

        assertNull(board!![INITIAL_POSITION])
        assertNotNull(board!!["A1"])
    }

    @Test
    fun getAvailablePositionsFromD5Test() = assertEquals(27, whiteQueen!!
            .moveTo("D5")
            .getAvailablePositions()
            .size)

    @Test
    fun getAvailablePositionsFromB4Test() = assertEquals(23, whiteQueen!!
            .moveTo(B1)
            .moveTo(B3)
            .getAvailablePositions()
            .size)

    @Test
    fun getAvailablePositionsFromA1Test() = assertEquals(21, whiteQueen!!
            .moveTo(A1)
            .getAvailablePositions()
            .size)

    @Test
    fun getAvailablePositionsFromH1Test() = assertEquals(21, whiteQueen!!
            .moveTo(H1)
            .getAvailablePositions()
            .size)

    @Test
    fun getAvailablePositionsFromA8Test() = assertEquals(21, whiteQueen!!
            .moveTo(A1)
            .moveTo(A8)
            .getAvailablePositions()
            .size)

    @Test
    fun getAvailablePositionsFromH8Test() = assertEquals(21, whiteQueen!!
            .moveTo(H1)
            .moveTo(H8)
            .getAvailablePositions()
            .size)
}